benchmark/tools/compare.py

```diff
@@ -0,0 +1,252 @@
+    return baseline / comparison
+
+
+def compare_benchmark(baseline: dict, comparison: dict) -> dict:
+    """Compares a handful of keys and component breakdowns recursively, writing them with a suffix to the output"""
+    result = {}
+    for key in sorted_key_intersection(baseline, comparison):
+        if key == "components":
+            assert isinstance(baseline[key], dict)
+            assert isinstance(comparison[key], dict)
+            result[key + suffix] = {
+                sub_key: ratio(baseline[key][sub_key], comparison[key][sub_key])
+                for sub_key in baseline[key]
+            }
+        elif isinstance(baseline[key], dict):
+            result[key] = compare_benchmark(baseline[key], comparison[key])
+        elif key in comparison_keys:
+            result[key + suffix] = ratio(baseline[key], comparison[key])
+    return result
+
+
+def compare(baseline: dict, comparison: dict) -> dict:
+    """Compares a test case, keeping root-level values and recursing into benchmarks"""
+    result = {}
+    for key in sorted_key_intersection(baseline, comparison):
+        # we don't have lists on the test case root level
+        assert not isinstance(baseline[key], list)
+        if isinstance(baseline[key], dict):
+            benchmark_result = {}
+            for benchmark_name in baseline[key].keys():
+                if isinstance(baseline[key][benchmark_name], dict):
+                    comparison_result = compare_benchmark(
+                        baseline[key][benchmark_name], comparison[key][benchmark_name]
+                    )
+                    if len(comparison_result) > 0:
+                        benchmark_result[benchmark_name] = comparison_result
+            if len(benchmark_result) > 0:
+                result[key] = benchmark_result
+        else:
+            # everything that's not a dict should only depend on the key in the root level
+            if baseline[key] != comparison[key]:
+                print(
+                    f"WARNING: Inconsistent value for {key}: {baseline[key]} != {comparison[key]}",
+                    file=sys.stderr,
+                )
+            result[key] = baseline[key]
+    return result
+
+
+def extract_benchmark_results(
+    input: dict, benchmarks: dict, case_key: tuple, context: str | None
+) -> None:
+    for key, value in input.items():
+        benchmark_name = key if context is None else f"{context}/{key}"
+        if key in map(lambda x: x + suffix, comparison_keys):
+            benchmark_name = benchmark_name[: -len(suffix)]
+            if benchmark_name not in benchmarks.keys():
+                benchmarks[benchmark_name] = []
+            benchmarks[benchmark_name].append((case_key, value))
+        elif isinstance(value, dict):
+            extract_benchmark_results(value, benchmarks, case_key, benchmark_name)
+
+
+def is_outlier(value: float, args) -> bool:
+    """returns true iff the value exceeds 1.0 above the outlier threshold"""
```

Comment:
Iff that is on purpose keep it, otherwise change it.

Reply:
iff = if and only if